crates/bevy_reflect/src/func/mod.rs

```diff
@@ -0,0 +1,220 @@
+//! Reflection-based dynamic functions.
+//!
+//! This module provides a way to pass around and call functions dynamically
+//! using the [`Function`] type.
+//!
+//! Many simple functions and closures can be automatically converted to [`Function`]
+//! using the [`IntoFunction`] trait.
+//!
+//! Once the [`Function`] is created, it can be called with a set of arguments provided
+//! via an [`ArgList`].
+//!
+//! This returns a [`FunctionResult`] containing the [`Return`] value,
+//! which can be used to extract a [`Reflect`] trait object.
+//!
+//!
+//! # Example
+//!
+//! ```
+//! # use bevy_reflect::func::args::ArgList;
+//! # use bevy_reflect::func::{Function, FunctionResult, IntoFunction, Return};
+//! fn add(a: i32, b: i32) -> i32 {
+//!   a + b
+//! }
+//!
+//! let mut func: Function = add.into_function();
+//! let args: ArgList = ArgList::default().push_owned(25_i32).push_owned(75_i32);
+//! let result: FunctionResult = func.call(args);
+//! let value: Return = result.unwrap();
+//! assert_eq!(value.unwrap_owned().downcast_ref::<i32>(), Some(&100));
+//! ```
+//!
+//! [`Reflect`]: crate::Reflect
+
+pub use error::*;
+pub use function::*;
+pub use info::*;
+pub use into_function::*;
+pub use return_type::*;
+
+pub use args::{Arg, ArgError, ArgList};
+
+pub mod args;
+mod error;
+mod function;
+mod info;
+mod into_function;
+pub(crate) mod macros;
+mod return_type;
+
+#[cfg(test)]
+mod tests {
```

Comment:
It would be nice to see tests on the limitations you provided, with `#[should_fail]`
- returns a reference tied to a different lifetime than the first argument
- type doesn't implement Reflect or FromArgs

Reply:
Yeah that would be good. I'll add some tests to `bevy_reflect_compile_fail`.